Correctly handle error column when coming from build log

vigoux opened the pull request:
Attempts to fix #738

Using the error message hint, this should not be too difficult to do,
at least in some cases.

vigoux commented:
@pfoerster I added the column field but I am not sure of my implementation to get the column out of the field. Furthermore, I am not sure how to test all this, is there a place where I can assert that the error markers are correctly placed by the LSP ?

pfoerster commented:
@vigoux Thanks for tackling this!

> I added the column field but I am not sure of my implementation to get the column out of the field.

I think the easiest way is to just track the `hint` in the `BuildError` and convert it to a column number at a later stage. This has the advantage that you do not need to do IO in the parser.

You can have a look at `src/diagnostics/build.rs`, which converts them to LSP diagnostics. Here, you have access to the `workspace`, which you can use to retrieve the source code of the document being edited.

> is there a place where I can assert that the error markers are correctly placed by the LSP

You can extend this integration test file: `tests/integration/lsp/text_document/publish_diagnostics.rs`. The tests spin up an actual server in memory.

vigoux commented:
Making some slow progress: I had to change the `&Workspace` to a `&mut Workspace` in some places, as now, collecting the build logs for a path may load a file.

Is that okay ?

vigoux commented:
Hi again.

How could I get the logging output of the LSP when running the tests ? It would be useful to debug this feature...

vigoux commented:
I think this is pretty much ready. I ended up being able to debug the issues that were present, and the change seems pretty much well tested. I am going to use this PR today quite extensively, and I'll see if I encounter anything weird.

pfoerster commented on src/diagnostics/build.rs:
```diff
@@ -22,21 +22,59 @@ pub fn collect_build_diagnostics(
+    let build_log_uri = Arc::clone(&build_log_document.uri);
 
-    let base_path = PathBuf::from(root_document.uri.path());
+    let mut base_path = PathBuf::from(root_document_uri.path());
```
You can use `root_document_uri.to_file_path()` here. The other method could cause problems on Windows.

pfoerster commented on src/diagnostics/build.rs:
```diff
@@ -9,11 +9,11 @@ use super::{Diagnostic, DiagnosticCode};
 pub fn collect_build_diagnostics(
     all_diagnostics: &DashMap<Arc<Url>, Vec<Diagnostic>>,
-    workspace: &Workspace,
+    workspace: &mut Workspace,
```
Mutating is no problem here, but note that the changes made to the workspace are not persisted. The general idea is that the `Workspace` on the main thread (the one that dispatches the LSP messages) is mutable. Change notifications are handled synchronously, while the LSP features are handled asynchronously so each request gets its own `Workspace` copy (which is relatively cheap to do). This way, requests that take longer to execute do not block the main thread.

pfoerster commented on src/diagnostics/build.rs:
```diff
@@ -22,21 +22,59 @@ pub fn collect_build_diagnostics(
+            if let Some(hint_line) = doc.text.lines().nth(line as usize) {
+                hint_line
+                    .find(hint)
+                    .map(|col| Position::new(line, (col + hint.len() - 1) as u32))
```
If we want to be sure, we need to convert the column offset using `doc.line_index` first. LSP uses `UTF-16` column offsets, while Rust uses `UTF-8` internally.

As an optimization, one could also use the `line_index` to speed up the lookup of the line, but I don't think this is really necessary here.

pfoerster reviewed (commented):
Looking good so far!

Regarding the logging statements in tests: 
I have added a small commit to the master branch. When you rebase your branch, you can use something like this to view the log output of a particular test.

`RUST_LOG=trace cargo t -- <name of the test case> --nocapture`

vigoux commented on src/diagnostics/build.rs:
```diff
@@ -22,21 +22,59 @@ pub fn collect_build_diagnostics(
+        let full_path = base_path.join(&error.relative_path);
+
+        let doc = if error.line.is_some() && error.hint.is_some() {
+            workspace.load(full_path.clone()).unwrap_or(None)
```
Nope, but that seemed to be the best solution.
It might be worth to expose a way to "fetch" a document, and return `None` if not present, I'll add that.

vigoux replied:
I switched to your method and that avoided to make `workspace` mutable, which is nicer for the diff.

vigoux commented:
So I _should_ have addressed all your comments now. Tell me if the fixes I did are okay from your POV.

Removing the `mut` for `workspace` allowed to remove most of the `Arc::clone` I introduced, so I'd expect better performance and memory usage too.

pfoerster reviewed (approved):
I have tried your changes out and it works very nice :+1
Thanks for contributing!
